roachtest: mark polled VM preemptions as non reportable

DarrylWong opened the pull request:
Previously, polled VM preemptions would simply cancel the test, as post test processing would recheck for preemptions again. However, we've seen some cases in AWS where the post test check returns no preemptions despite the polling returning preemptions.

This may be just be the AWS check being eventually consistent, so we want to avoid posting if either check finds preemptions.

----

The second change resets failures in the case of a vm preemption, in case a timeout occurred which normally takes precedence over all other failures. While a timeout suggests that something should be fixed with the test (usually respecting the test context cancellation), we see that in practice, engineers tend to close the issue without investigating as soon as they see the preemption.

This also removes the potential duplicate vm_preemption failure that may have been added by the preemption polling.

Fixes: https://github.com/cockroachdb/cockroach/issues/139004
Fixes: https://github.com/cockroachdb/cockroach/issues/139931
Release note: none
Epic: none

DarrylWong committed:
roachtest: mark polled VM preemptions as non reportable

Previously, polled VM preemptions would simply cancel the
test, as post test processing would recheck for preemptions
again. However, we've seen some cases in AWS where the post test
check returns no preemptions despite the polling returning
preemptions.

This may be just be the AWS check being eventually consistent,
so we want to avoid posting if either check finds preemptions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139075)
<!-- Reviewable:end -->

DarrylWong committed:
roachtest: reset failures on vm preemption

This change resets failures in the case of a vm preemption,
in case a timeout occurred which normally takes precedence
over all other failures. While a timeout suggests that
something should be fixed with the test (usually respecting
the test context cancellation), we see that in practice,
engineers tend to close the issue without investigating as
soon as they see the preemption.

This also removes the potential duplicate vm_preemption
failure that may have been added by the preemption polling.

DarrylWong commented:
On the fence on the second commit. One one hand there have been two cases where teams fix their tests, but in general it appears [most engineers just close out the issue](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue%20state%3Aclosed%20%22VMs%20preempted%20during%20the%20test%20run%3A%22). Arguable how much value is even gained from fixing said tests anyway.

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,54 @@ func TestVMPreemptionPolling(t *testing.T) {
+	// Test that if VM preemption polling finds a preempted VM but the post test failure
+	// check doesn't, the test is still marked as a flake.
+	t.Run("post test check doesn't catch preemption", func(t *testing.T) {
+		setPollPreemptionInterval(50 * time.Millisecond)
```
This would flake under `--stress`, right?

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,54 @@ func TestVMPreemptionPolling(t *testing.T) {
+		}
+		mockTest.Timeout = 50 * time.Millisecond
+
+		err := runner.Run(ctx, []registry.TestSpec{mockTest}, 1, /* count */
```
Same as above...  timeout may never occur under `--stress`.

DarrylWong replied:
Lowered the timeout to 10 ms, we have another unit test that uses that time and seems to be doing fine, so hopefully it should be for this as well. Stressed 3000 times with no flakes.

DarrylWong commented:
Yup, `stress` didn't fail but `race` did. Interestingly, the failure being added by the preemption polling races with the post test assertion check replacing the logger. This isn't specific to preemption polling, I can repro this by just running:

```
		mockTest.Run = func(ctx context.Context, t test.Test, c cluster.Cluster) {
			t.Error("test failed")
		}
		mockTest.Timeout = 1 * time.Nanosecond
```

So this is a possible data race we could see in actual roachtests if it times out. From what I can tell, the only impact is that we might lose the failure being logged to `test.log`, which is not a huge deal since the test timed out which is most likely the real failure anyway.

For these unit tests, I just made sure the above race doesn't happen since it doesn't affect what it's testing. Not sure if we want to add a real fix for the data race though.

srosenberg commented:
>  Interestingly, the failure being added by the preemption polling races with the post test assertion check replacing the logger.

Not entirely following... I thought `addFailure` was thread-safe. Can you post the report from the data race, i.e., the racing goroutine stacks.

DarrylWong commented:
Sure, here's a simple repro of what I'm seeing:

```
func TestRunnerTestTimeoutRace(t *testing.T) {
	ctx := context.Background()
	stopper := stop.NewStopper()
	defer stopper.Stop(ctx)
	cr := newClusterRegistry()
	runner := newUnitTestRunner(cr, stopper)

	var buf syncedBuffer
	copt := defaultClusterOpt()
	lopt := defaultLoggingOpt(&buf)
	test := registry.TestSpec{
		Name:             `timeout`,
		Owner:            OwnerUnitTest,
		Timeout:          1 * time.Millisecond,
		Cluster:          spec.MakeClusterSpec(0),
		CompatibleClouds: registry.AllExceptAWS,
		Suites:           registry.Suites(registry.Nightly),
		CockroachBinary:  registry.StandardCockroach,
		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
			t.Error("test failed")
		},
	}
	_ = runner.Run(ctx, []registry.TestSpec{test}, 1, /* count */
		defaultParallelism, copt, testOpts{}, lopt)
}
```

Running `./dev test pkg/cmd/roachtest -f TestRunnerTestTimeoutRace --stress --race` hits:

```
==================
WARNING: DATA RACE
Read at 0x00c004b9e888 by goroutine 40:
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testImpl).L()
      pkg/cmd/roachtest/test_impl.go:262 +0x520
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testImpl).addFailure()
      pkg/cmd/roachtest/test_impl.go:470 +0x530
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testImpl).addFailureAndCancel()
      pkg/cmd/roachtest/test_impl.go:441 +0x5c
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testImpl).Error()
      pkg/cmd/roachtest/test_impl.go:432 +0x4c
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.TestRunnerTestTimeoutRace.func1()
      pkg/cmd/roachtest/test_test.go:381 +0x88
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runTest.func2()
      pkg/cmd/roachtest/test_runner.go:1315 +0x168

Previous write at 0x00c004b9e888 by goroutine 39:
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testImpl).ReplaceL()
      pkg/cmd/roachtest/test_impl.go:268 +0x1d0
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runTest.func4()
      pkg/cmd/roachtest/test_runner.go:1380 +0x1cc
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runTest()
      pkg/cmd/roachtest/test_runner.go:1402 +0x14d8
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runWorker()
      pkg/cmd/roachtest/test_runner.go:937 +0x2b6c
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).Run.func1()
      pkg/cmd/roachtest/test_runner.go:389 +0x3b4
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      pkg/util/stop/stopper.go:500 +0x214
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.gowrap1()
      pkg/util/stop/stopper.go:501 +0x54

Goroutine 40 (running) created at:
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runTest()
      pkg/cmd/roachtest/test_runner.go:1296 +0xba8
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).runWorker()
      pkg/cmd/roachtest/test_runner.go:937 +0x2b6c
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).Run.func1()
      pkg/cmd/roachtest/test_runner.go:389 +0x3b4
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
      pkg/util/stop/stopper.go:500 +0x214
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.gowrap1()
      pkg/util/stop/stopper.go:501 +0x54

Goroutine 39 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      pkg/util/stop/stopper.go:489 +0x52c
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      pkg/util/stop/stopper.go:360 +0xfd4
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.(*testRunner).Run()
      pkg/cmd/roachtest/test_runner.go:378 +0xcc4
  github.com/cockroachdb/cockroach/pkg/cmd/roachtest.TestRunnerTestTimeoutRace()
      pkg/cmd/roachtest/test_test.go:384 +0x5e4
  testing.tRunner()
      GOROOT/src/testing/testing.go:1689 +0x180
  testing.(*T).Run.gowrap1()
      GOROOT/src/testing/testing.go:1742 +0x40
==================
```

So the `failures` struct in `addFailure` is thread safe, but the logger it uses here: 

https://github.com/cockroachdb/cockroach/blob/1de6c82abe9e266af65bcd9428548aad8d9abd33/pkg/cmd/roachtest/test_impl.go#L349

isn't because we replace the logger after the test is done:

https://github.com/cockroachdb/cockroach/blob/3c64af25d55d8ce568e5d9735ee74a87b9d3c32e/pkg/cmd/roachtest/test_runner.go#L1090-L1098

Note the comment: `We still want to run the post-test assertions even if the test timed out` along with the fact that on test timeouts we purposely don't cancel the test context and let the test keep running.

So my understanding is that the following ordering can happen:

1. The test times out and the test runner stops blocking.
2. The test runner starts post tests assertion checks, replacing the logger.
3. The test itself runs `t.Error("test failed")` which eventually calls `t.L().Printf`. This races with the above.

----

The original test was hitting the same race w/ the same stack trace, it was just the polling preemption adding the failure instead of the test itself.

srosenberg commented:
> The test times out and the test runner stops blocking.
The test runner starts post tests assertion checks, replacing the logger.
The test itself runs t.Error("test failed") which eventually calls t.L().Printf. This races with the above.

Yep. That's a perfectly plausible scenario. We should fix it; e.g., changing the type of `testImpl.L` to `atomic.Pointer`.

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,62 @@ func TestVMPreemptionPolling(t *testing.T) {
+					return nil, nil
+				}
+			}()
+			// Make sure the preemption polling is called and the test context is cancelled
```
Good. This removes the clock, and the corresponding "flakiness", from the equation.

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,62 @@ func TestVMPreemptionPolling(t *testing.T) {
+				<-ctx.Done()
+			}
+		}
+		mockTest.Timeout = 10 * time.Millisecond
```
Nit: I'd move this before `Run` since, logically, `Timeout` is intended to modify `Run's` behavior.

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,62 @@ func TestVMPreemptionPolling(t *testing.T) {
+		}
+
+		mockTest.Run = func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			for {
```
Nit: `for` is redundant since a cancelled context remains cancelled.

DarrylWong replied:
Ah true, changed. I wanted it to block forever but linter was complaining about an empty for loop. Blocking on the context does what I want though.

srosenberg commented on pkg/cmd/roachtest/test_test.go:
```diff
@@ -767,4 +767,62 @@ func TestVMPreemptionPolling(t *testing.T) {
+			return preemptedVMs, nil
+		}
+
+		mockTest.Run = func(ctx context.Context, t test.Test, c cluster.Cluster) {
```
We should add an explainer since it's not immediately clear as to what's expected to happen. We expect that the context will be cancelled only when `Timeout` expires, at which point the test would normally be marked as failed. However, due to injected preemption, the test isn't marked as failed. Thus, `runner.Run` is expected not to return `errTestsFailed`, or any other error.

DarrylWong replied:
Done.

DarrylWong commented:
> We should fix it; e.g., changing the type of testImpl.L to atomic.Pointer

SGTM, will put it out as a followup PR.

srosenberg commented:
blathers backport 25.1.0-rc
